compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollJamAIEVecMatmul.cpp

```diff
@@ -0,0 +1,448 @@
+      if (inType.isBF16()) {
+        // Where is this sequence from?
+        // I did some manual tuning by running the test
+        // run.py with '--tests=CorePerformance' of different sequences.
+        // I tried the following sequences:
+        //   uj_1_4_u_0_4_u_0_2_u_0_2
+        //   uj_1_4_uj_2_2_u_0_4
+        //   uj_1_4_uj_2_4_u_0_4
+        //   uj_2_4_uj_1_4_u_0_4
+        //   uj_1_2_uj_2_2_u_0_4_u_0_4
+        //   uj_0_4_uj_1_2_uj_2_2_u_0_4
+        //   uj_0_2_uj_2_2_uj_1_2_u_0_2_u_0_4
+        //   uj_0_4_uj_0_2_uj_1_2_u_0_2
+        //   uj_1_2_uj_0_2_uj_2_2_u_0_2_u_0_4
+        // Some were very slow, some were reasonable, some resulted in stack
+        // overflows.
+        if (loopCounts == SmallVector<int>{4, 8, 8}) {
+          // this sequence of transforms would be encoded by the string:
+          // "uj_0_2_uj_2_2_u_0_2_u_0_4_u_0_2_NOUNROLL";
+          transformations.push_back({UnrollType::UnrollJam, 0, 2});
+          transformations.push_back({UnrollType::UnrollJam, 2, 2});
+          transformations.push_back({UnrollType::Unroll, 0, 2});
+          transformations.push_back({UnrollType::Unroll, 0, 4});
+          transformations.push_back({UnrollType::Unroll, 0, 2});
+          noUnrollAtEnd = true;
+          return success();
+        }
+      }
+    }
+
+    return success();
+  }
+
+  // Initialize \p transformations and \p noUnrollAtEnd from the string
+  // \p sequence. If the sequence is invalid, return a failure.
+  static LogicalResult initialize(
+      aievec::MatMulOp matmul, std::string_view sequence,
+      SmallVector<std::tuple<UnrollType, int, int>> &transformations,
+      bool &noUnrollAtEnd) {
+    transformations.clear();
+    noUnrollAtEnd = false;
+
+    if (sequence.empty() || sequence == "none") return success();
+
+    if (sequence == "default") {
+      return initializeTuned(matmul, transformations, noUnrollAtEnd);
+    }
+
+    SmallVector<StringRef> splits;
+    llvm::StringRef(sequence).split(splits, '_', /*MaxSplit=*/-1,
+                                    /*KeepEmpty=*/false);
+
+    auto errStart = [&]() -> std::string {
+      return std::string("has an unroll sequence \"") + std::string(sequence) +
+             "\" ";
+    };
+
+    int splitsMod = splits.size() % 3;
+
+    // Handle the case where the sequence ends with "NOUNROLL" or "UNROLL".
+    if (splitsMod == 1) {
+      llvm::StringRef noUnrollAtEndStr = splits.back();
+      if (noUnrollAtEndStr == "NOUNROLL")
+        noUnrollAtEnd = true;
+      else if (noUnrollAtEndStr == "UNROLL")
+        noUnrollAtEnd = false;
+      else {
+        return matmul->emitOpError()
+               << errStart() << "with an unknown unroll/no-unroll at end, '"
+               << noUnrollAtEndStr << "'. Expected 'NOUNROLL' or 'UNROLL'.";
+      }
+      splits.pop_back();
+    }
+
+    else if (splitsMod == 2) {
+      return matmul->emitOpError()
+             << errStart() << "whose length is is 3*n + 2 for some n.";
```

Comment:
```suggestion
             << errStart() << "whose length is 3*n + 2 for some n.";
```
Also I'm not quite understand what this error means.

Reply:
Remainder when divided by 3 is 2. There are 2 supported formats. If remainder is 0, it's something like `uj_0_2_u_1_4`. If remainder is 1, it's like the remainder 0 case but with `_UNROLL` or `_NOUNROLL` in the end. If remainder is 2: invalid.